### DISCUSSION TITLE
Best way to combine a and sft? And are some functions redundant?

### DISCUSSION THREAD

**Discussion by txtyash** [kmonad/kmonad#406] [Ideas]
`as (tap-hold 153 a lsft)` is what I have in my config.

Is this the best way to do it?

I know there are other functions like next, release, and their combinations with each other but the kmonad reference is not verbose enough, so I don't know whether they're useful.

The only caveat I see with the binding I use is that it has an input lag of 153ms. Do other functions mentioned above have the same downside?

**Comment by MuhammedZakir** [comment]
> I know there are other functions like next, release, and their combinations with each other but the kmonad reference is not verbose enough, so I don't know whether they're useful.

Have you read both `doc/quick-reference.md` and `keymap/tutorial.kbd`, or atleast the latter? Also, the best way is to try them out yourself.

> The only caveat I see with the binding I use is that it has an input lag of 153ms.

Why not use a lesser value then?

> Do other functions mentioned above have the same downside?

It depends on your usecase and typing habit. As I said above, I recommend you to try those buttons yourself.

**Comment by txtyash** [comment]
I've read all of the reference and I've read most of the docs but I just wanted to reduce the input lag. If all of them are going to have some input lag then there's no point in trying out all of them. What I want is  that If I press and release the `a` key, no matter how long it was being held down then it should be registered as `a` else if I press the `a` key and then press another key while the `a` key is held down then it should be registered as `sft`. This solves my problem of input lag. Can this be done?

**Reply by ibnuda** [reply]
one thing though. are you sure you can tap both key at exactly the same time without any delay (not even 1 millisecond) consistently?

**Reply by txtyash** [reply]
> one thing though. are you sure you can tap both key at exactly the same time without any delay (not even 1 millisecond) consistently?

No, obviously not. I'll explain my problem in more detail:

Suppose I have mapped the `a` key to do `a` when tapped and 'sft' when held. It has a timeout of 150ms. 

Now lets get to typing. I average at around 85wpm and definitely notice the lag when I type `a` which would look acceptable to you but not to me. So now I would think of reducing it even more but then I would sometimes miss `a` as it would be registered as shift if I'm not fast enough.

Input lag is always a terrible experience. I would want it to be more when I'm casually typing and I would want it to be much less when I'm typing fast but not too less to miss the key-stroke entirely.

I believe that this can be solved and the input lag due to the timeout can be completely eliminated. I don't know if there are already any functions available to do this. I think there aren't any because I've read the docs and gone through all of the user configs including  @david-janssen's

Now to eliminate input lag for the `a` key what we can do is have a function which registers the `a` key as `a` when it is pressed and released no matter how quick or how slow, so 0 input lag as no timeout but if I'd pressed another key while `a` was held down then it will be registered as `shift`.

I would like to draw @slotThe's and @david-janssen's attention to this.

**Reply by ibnuda** [reply]
basically you want qmk's `NO_PERMISSIVE_HOLD`, then.

**Reply by txtyash** [reply]
> basically you want qmk's `NO_PERMISSIVE_HOLD`, then.

ugh... I can't use qmk on my laptop.

**Reply by txtyash** [reply]
> basically you want qmk's `NO_PERMISSIVE_HOLD`, then.

Is something similar available in kmonad?

**Reply by txtyash** [reply]
> basically you want qmk's `NO_PERMISSIVE_HOLD`, then.

Or maybe another linux tool that has this ability?

**Comment by m-althaf** [comment]
> `as (tap-hold 153 a lsft)` is what I have in my config.
> 
> Is this the best way to do it?
> 
> I know there are other functions like next, release, and their combinations with each other but the kmonad reference is not verbose enough, so I don't know whether they're useful.
> 
> The only caveat I see with the binding I use is that it has an input lag of 153ms. Do other functions mentioned above have the same downside?

from what l can understand `tap-hold` isn't the button for shifting maybe try `tap-next` (remember you lose the key repeat) or other `tap-next....`

**Reply by txtyash** [reply]
> > `as (tap-hold 153 a lsft)` is what I have in my config.
> > Is this the best way to do it?
> > I know there are other functions like next, release, and their combinations with each other but the kmonad reference is not verbose enough, so I don't know whether they're useful.
> > The only caveat I see with the binding I use is that it has an input lag of 153ms. Do other functions mentioned above have the same downside?
> 
> from what l can understand `tap-hold` isn't the button for shifting maybe try `tap-next` (remember you lose the key repeat) or other `tap-next....`

I'd tried it before and tried it now, even as I am typing and I can tell you that it is terrible. This has so much lag and space key doesn't even get registered most of the times while quickly typing.

Tried it with this binding:
`custom (tap-next spc lmet)`

**Reply by txtyash** [reply]
Hey @Ultrahalf,
Can you take a look at https://github.com/kmonad/kmonad/issues/407

**Comment by gerardbosch** [comment]
Hi @zim0369 I think `tap-hold` is not what you're looking for.

Here is what I understood these functions do:
```lisp
;;               `tap-hold`  emits ONLY AFTER the timeout (not useful for regular mod keys). If the key is released
;;                             before timeout expiration, the key along with any subsequent keys are emited.
;;          `tap-hold-next`  emits AS SOON AS a 2nd key is pressed. The timeout tells when the key becomes a modifier
;;                             if no 2nd key is pressed yet.
```
**I think what you're looking for is `tap-hold-next-release`** :smile: which I think is the most comfortable to use for home-row mods (that's what I'm using and what I saw from other people). You can check [my configuration (WIP)](https://github.com/gerardbosch/dotfiles/blob/master/home/.config/kmonad/mi-keyboard.kbd) to see if it helps you. I'm using dual keys for <kbd>Ctrl</kbd>/<kbd>Sft</kbd> on <kbd>A</kbd>/<kbd>S</kbd> [left] and <kbd>;</kbd>/<kbd>L</kbd> [right], and other stuff.

When I want to write `Hello`, I get the `H` as soon as I press `h` (while holding `s`), no delay involved.
It seemed to me that the difference with regular shift button is that regular shift seems to be sticky, which helps in fast typing, but in the dual-key case, you need to be sure to hold the mod to emit the uppercase character.

---
Also, may you interest to check the `:timeout-button` option in case you want to restore the repeating key behavior:
`lSft (tap-hold-next-release 250 l sft :timeout-button l)`. DISCLAIMER: This option is not shipped at the current release v0.4.1, so you need to build KMonad from source to make it work.
